### ISSUE TITLE
Docker rm -f has a race condition surrounding timing with the Soft Power Off

### ISSUE THREAD

**Issue by matthewavery** [vmware/vic#6252]
**Details:**
`docker rm -f` fails with `cannot remove a running container`. On the second run the container is removed successfully without issue. This appears to be a state race. We are initiating a hard power off when we perceive the soft power off has failed. However, the soft power off does succeed, because of this we actually get an error for the hard power off which causes us to error out. However, this error is that the container is already off(which is what we wanted anyways). so instead of failing we should continue with the removal. 

note: a simple workaround is to just run the `docker rm -f` command again(no modification needed). 

**VIC version:**

VIC ENGINE : 1.2_RC5 --> VIC-OVA-RC3   

**Deployment details:**

Any deployed VCH which can create and destroy containers is vulnerable to this bug. It is much easier to reproduce if you perform concurrent creates and rms on 10+ containers. 

**Steps to reproduce:**

customer ran into this by doing 

docker create <image> * 7 in parallel
docker rm -f <cid> * 7 in parallel

**Actual behavior:**

docker rm -f <cid> --> "cannot remove a running container"

**Expected behavior:**

docker rm -f <cid> --> "<cid>" 

**Logs:**

we cannot provide logs for this from the repro with the customer since they reveal some aspects of the customers environment. Reproduction of this issue should not be too terribly difficult, but the state race is somewhat straight forward in analysis.

**Comment by matthewavery** [comment]
For Posterity these are the notes taken onsite during the advent of the bug.

```
For the failure. This was a standard state race. Powering off a container on the hard poweroff. The hard poweroff can fail if... on the soft power off we are successful but it takes long enough that we attempt the hard poweroff. Then during the attempted hard poweroff we get an error stating "the container is already off". But... the error case for this in the code reports "cannot turn off a running container"  in the event of an error. This is a bug in VIC. What we should do is note that the container did reach the target state and then proceed with the removal of the container. Right now the implication is that the rm command will need to be run a second time.
```

**Comment by matthewavery** [comment]
I will be dropping the analysis in two places. Here and in #6236 since both of these issues are one and the same. I will be using logging from #6236 to break down what is happening.

**Comment by matthewavery** [comment]
There are two primary issues encompassed in this scenario. One we are not fully seeing the effect of from the docker cli, but it is present in the logs. it is this log message

**Comment by matthewavery** [comment]
We have found significant evidence of this issue in the customers logs. @mdubya66 Especially when deleting large batches(I think at one point it was 80 - 100). It has not been hard for us to repro it in the customers environement.

**Comment by chengwang86** [comment]
@matthewavery For number two, in c.poweroff(), we return `nil` even if we get `invalidpowerstate` error. So this error should not fail `docker stop` or `docker rm -f`.

**Comment by mdubya66** [comment]
putting into 1.3 and marking high pri, per @pdaigle

**Comment by sflxn** [comment]
I believe https://github.com/vmware/vic/pull/6254 resolved this issue.  Whoever takes it next should verify.

**Comment by hickeng** [comment]
@sflxn You opened #6672 which would be likely to recreate this if it is still an issue. Did you observe this problem in the #6672 logs? There's no symptoms noted in that issue.

**Comment by sflxn** [comment]
@hickeng Yes, I did see "cannot remove a running container" quite a bit in #6672.  Not sure if it was directly related to this issue.  However, that issue was written so someone could actually look at the logs and identify all the issue that was found.  There were at least 3 issues found from a scan of the logs.

**Comment by sflxn** [comment]
@hickeng So I updated #6672 with some new scripts and a log, parallel_10_improved.log that capture an instance of this error, "cannot remove a running container".

**Comment by matthewavery** [comment]
also see #6263 this has a fresh nightly failure as of 11/7/2017. reference logs there

**Comment by sflxn** [comment]
I just looked at the logs found in #6263.  It's the same original problem with docker stop from two months ago.  I did an analysis of this same issue about two months ago, and it's the same analysis.  The solution that was added (not my solution) during that time was to attempt retry, but it obviously didn't fix it completely.

The problem starts when execute a stop() in commit.go.  That's a synchronous call that successfully does stop the container VM, but as Matt mentions above, there's some issues during this operation.  Nevertheless, it succeeds, the container is stopped, we send out an event that the container is stopped.  The problem is the next line, where we perform refresh() on the handle and hit the host sync problem.  We query vSphere for the power state, but vSphere returns false power state.  Two months ago, when I analyzed this problem, the hostd file showed that vSphere eventually sends the correct update, presumably when it gets a host sync.

Then we hit some code that was added 3 months ago to speculatively check for concurrent access.  It's now using stale power state from the refresh() and then incorrectly returns concurrent access error.

Tomorrow, I'll look at the hostd file from the recent logs I looked at and post the evidence.  This should not be a difficult fix.

**Comment by sflxn** [comment]
```
2017-11-07T09:54:01.389Z info hostd[41880B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx] State Transition (VM_STATE_ON -> VM_STATE_OFF)
2017-11-07T09:54:01.391Z info hostd[42B80B70] [Originator@6876 sub=Guestsvc.GuestFileTransferImpl] Entered VmPowerStateListener
2017-11-07T09:54:01.391Z info hostd[42B80B70] [Originator@6876 sub=Guestsvc.GuestFileTransferImpl] VmPowerStateListener succeeded
2017-11-07T09:54:01.391Z info hostd[42B80B70] [Originator@6876 sub=Hbrsvc] Replicator: powerstate change VM: 1 Old: 1 New: 0
2017-11-07T09:54:01.391Z info hostd[42B80B70] [Originator@6876 sub=Hbrsvc] Replicator: Poweroff for VM: (id=1)
2017-11-07T09:54:01.391Z warning hostd[41880B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx] Failed to find activation record, event user unknown.

event manager noticed VM is powered off
---------------------------------------
2017-11-07T09:54:01.391Z info hostd[41880B70] [Originator@6876 sub=Vimsvc.ha-eventmgr] Event 245 : romantic_knuth-e633100ac9c0 on  sc-rdops-vm12-dhcp-103-81.eng.vmware.com in ha-datacenter is powered off


[LikewiseGetDomainJoinInfo:355] QueryInformation(): ERROR_FILE_NOT_FOUND (2/0):
[LikewiseGetDomainJoinInfo:355] QueryInformation(): ERROR_FILE_NOT_FOUND (2/0):


Attempt to hard power off?
--------------------------
2017-11-07T09:54:10.314Z info hostd[41A8CB70] [Originator@6876 sub=Vimsvc.TaskManager opID=5531c40c-9e-ff4b user=vpxuser:VSPHERE.LOCAL\Administrator] Task Created : haTask-1-vim.VirtualMachine.powerOff-874
2017-11-07T09:54:10.316Z info hostd[42580B70] [Originator@6876 sub=Vimsvc.ha-eventmgr opID=5531c40c-9e-ff4b user=vpxuser:VSPHERE.LOCAL\Administrator] Event 246 : romantic_knuth-e633100ac9c0 on  sc-rdops-vm12-dhcp-103-81.eng.vmware.com in ha-datacenter is stopping


Invalid State transition
------------------------
2017-11-07T09:54:10.316Z error hostd[42580B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx opID=5531c40c-9e-ff4b user=vpxuser:VSPHERE.LOCAL\Administrator] Invalid transition requested (VM_STATE_OFF -> VM_STATE_POWERING_OFF): Invalid state
2017-11-07T09:54:10.317Z info hostd[42580B70] [Originator@6876 sub=Default opID=5531c40c-9e-ff4b user=vpxuser:VSPHERE.LOCAL\Administrator] AdapterServer caught exception: vim.fault.InvalidPowerState
2017-11-07T09:54:10.317Z info hostd[42580B70] [Originator@6876 sub=Vimsvc.TaskManager opID=5531c40c-9e-ff4b user=vpxuser:VSPHERE.LOCAL\Administrator] Task Completed : haTask-1-vim.VirtualMachine.powerOff-874 Status error
2017-11-07T09:54:10.322Z info hostd[41640B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] State Transition (VM_STATE_OFF -> VM_STATE_NAMESPACE_CHANGE)
2017-11-07T09:54:10.323Z error hostd[41640B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Namespace 'ovf' not found.
2017-11-07T09:54:10.324Z info hostd[41640B70] [Originator@6876 sub=Vmsvc.vm:/vmfs/volumes/cc8bbd3c-107b2b14/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662/e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662.vmx opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] State Transition (VM_STATE_NAMESPACE_CHANGE -> VM_STATE_OFF)
2017-11-07T09:54:10.325Z info hostd[41640B70] [Originator@6876 sub=Default opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] AdapterServer caught exception: vim.fault.NotFound
2017-11-07T09:54:10.325Z info hostd[41640B70] [Originator@6876 sub=Solo.Vmomi opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Activation [N5Vmomi10ActivationE:0x41df6520] : Invoke done [queryNamespace] on [vim.vm.NamespaceManager:ha-vm-namespace-manager]
2017-11-07T09:54:10.325Z verbose hostd[41640B70] [Originator@6876 sub=Solo.Vmomi opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Arg vm:
--> 'vim.VirtualMachine:1'
2017-11-07T09:54:10.325Z verbose hostd[41640B70] [Originator@6876 sub=Solo.Vmomi opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Arg name:
--> "ovf"


Fault thrown because of the power state exception
-------------------------------------------------
2017-11-07T09:54:10.325Z info hostd[41640B70] [Originator@6876 sub=Solo.Vmomi opID=5
531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Throw vim.fault.NotFound
2017-11-07T09:54:10.325Z info hostd[41640B70] [Originator@6876 sub=Solo.Vmomi opID=5531c40c-9e-ff51 user=vpxuser:VSPHERE.LOCAL\Administrator] Result:
--> (vim.fault.NotFound) {
-->    faultCause = (vmodl.MethodFault) null,
-->    msg = ""
--> }
```

The hostd file didn't show the same activity as the one from a few months ago.  It seems to have less logging so it's not as evident what's going on.  However, I am going to speculate that once vSphere detects the exception above and "throws" an error, the error handling bypasses normal processing, which later prevents VIC's refresh code from getting updated power state from a property query.

**Comment by sflxn** [comment]
From portlayer logs:

```
stop
----
Nov  7 2017 09:53:50.154Z DEBUG [BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*Container).stop:446] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:53:50.154Z DEBUG Setting container e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 state: Stopping
Nov  7 2017 09:53:50.154Z INFO  sending kill -TERM e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:53:50.154Z DEBUG [BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).startGuestProgram:187] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662:kill
Nov  7 2017 09:53:50.403Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).startGuestProgram:187] [249.302226ms] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662:kill
Nov  7 2017 09:53:50.403Z INFO  waiting 10s for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 to power off
Nov  7 2017 09:53:50.844Z DEBUG vSphere Event Guest operation Start Program performed on Virtual machine romantic_knuth-e633100ac9c0. for eventID(958) ignored by the event collector
Nov  7 2017 09:54:00.156Z WARN  timeout (10s) waiting for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 to power off via SIGTERM
Nov  7 2017 09:54:00.156Z INFO  sending kill -KILL e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:54:00.156Z DEBUG [BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).startGuestProgram:187] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662:kill
Nov  7 2017 09:54:00.287Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).startGuestProgram:187] [130.636724ms] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662:kill
Nov  7 2017 09:54:00.287Z INFO  waiting 10s for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 to power off
Nov  7 2017 09:54:00.841Z DEBUG vSphere Event Guest operation Start Program performed on Virtual machine romantic_knuth-e633100ac9c0. for eventID(959) ignored by the event collector
Nov  7 2017 09:54:10.165Z WARN  timeout (10s) waiting for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 to power off via SIGKILL

hard power off
--------------
Nov  7 2017 09:54:10.170Z WARN  stopping e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 via hard power off due to: failed to shutdown e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 via kill signals [TERM KILL]
Nov  7 2017 09:54:10.170Z DEBUG op=299.14 (delta:5.043µs): [NewOperation] op=299.14 (delta:1.929µs) [github.com/vmware/vic/pkg/vsphere/tasks.WaitForResult:68]

detected power off
------------------
Nov  7 2017 09:54:10.301Z ERROR op=299.14 (delta:131.163828ms): unexpected fault on task retry: &types.InvalidPowerState{InvalidState:types.InvalidState{VimFault:types.VimFault{MethodFault:types.MethodFault{FaultCause:(*types.LocalizedMethodFault)(nil), FaultMessage:[]types.LocalizableMessage{types.LocalizableMessage{DynamicData:types.DynamicData{}, Key:"vpxd.vm.poweroff.unexpectedfailure", Arg:[]types.KeyAnyValue{types.KeyAnyValue{DynamicData:types.DynamicData{}, Key:"1", Value:"romantic_knuth-e633100ac9c0"}}, Message:"An error was received from the ESX host while powering off VM romantic_knuth-e633100ac9c0."}}}}}, RequestedState:"poweredOn", ExistingState:"poweredOff"}

Return from base.go's poweroff where types.InvalidPowerState was detected
-------------------------------------------------------------------------------------------------------
Nov  7 2017 09:54:10.325Z WARN  power off e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 task skipped (state was already poweredOff)
Nov  7 2017 09:54:10.325Z DEBUG Set container e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 state: Stopped
Nov  7 2017 09:54:10.325Z DEBUG Container(e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662) closing 0 log followers

stop completed
--------------
Nov  7 2017 09:54:10.325Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*Container).stop:446] [20.171675025s] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662

refresh
-------
Nov  7 2017 09:54:10.325Z DEBUG [BEGIN] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).updates:121] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:54:10.325Z DEBUG Update: for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662, refreshing from change version 2017-11-07T09:50:35.609429Z
Nov  7 2017 09:54:10.325Z DEBUG get vm properties [config runtime] of vm VirtualMachine:vm-34
Nov  7 2017 09:54:10.325Z DEBUG properties: [config runtime summary.runtime.connectionState]
Nov  7 2017 09:54:10.356Z DEBUG Update: for e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662, change version 2017-11-07T09:50:35.609429Z, extraconfig id: e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:54:10.356Z DEBUG [BEGIN] [github.com/vmware/vic/lib/migration.migrateConfig:77] target: ContainerConfigure, version key: guestinfo.vice./version/PluginVersion
Nov  7 2017 09:54:10.356Z DEBUG [ END ] [github.com/vmware/vic/lib/migration.migrateConfig:77] [48.838µs] target: ContainerConfigure, version key: guestinfo.vice./version/PluginVersion
Nov  7 2017 09:54:10.356Z DEBUG [ END ] [github.com/vmware/vic/lib/portlayer/exec.(*containerBase).updates:121] [31.002312ms] e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662
Nov  7 2017 09:54:10.356Z INFO  Reconfigure: attempting update to e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 with change version "2017-11-07T09:50:35.609429Z" (poweredOn)

sending out concurrent access error
-----------------------------------
Nov  7 2017 09:54:10.356Z WARN  Reconfigure: collision of concurrent operations - expected power state poweredOff, found poweredOn
Nov  7 2017 09:54:10.428Z DEBUG Container e633100ac9c0bd6218d286ef0acbd14e991bfd2deb1fc11e76a4684dd9a77662 power state: poweredOn, runtime power state: poweredOn
```

What the above snippet shows is that we can actually detect power state was already off.  Instead of returning nil, we can return something else so that commit.go can then check against the refresh power state and determine the refresh didn't get updated state.

**Comment by sflxn** [comment]
During my test of my potential fix, I found that there is more than one condition that causes this stale power state to occur.  We've mostly seen the case where the container is already powered off and we go to perform a hard shutdown.

The other case is when the guest is shutting down.  I haven't seen this show up in nightly tests or in our repo, but my concurrent script in a nested environment can reproduce this.  This case:
https://github.com/vmware/vic/blob/master/lib/portlayer/exec/base.go#L366

**Comment by sflxn** [comment]
Scenario 1:
---------------
- soft power off
- VM powered off but VIC doesn't notice
- hard power identifies this error and return nil
- commit() tries to refresh and gets stale power state

scenario 2:
--------------
- soft power off
- guest in process of powering off
- hard power
- hard power attempt causes vSphere exception handling
- hard power off identifies guest is powering off and returns nil
- commit() refreshes properties and get stale power state

scenario 3:
--------------
- soft power off
- vm powers off
- VIC doesn't think it has and performs another soft power
- vSphere exception occurs
- soft power returns nil
- commit() refreshes properties and get stale power state

**Comment by sflxn** [comment]
Here are some logs from scenario 3.  These are logs from my concurrent script, container log bundle, and portlayer logs marked up with my commentary.

[container-logs.zip](https://github.com/vmware/vic/files/1472115/container-logs.zip)
[parallel_30_both.log](https://github.com/vmware/vic/files/1472118/parallel_30_both.log)
[port-layer_both.log](https://github.com/vmware/vic/files/1472120/port-layer_both.log)

**Comment by sflxn** [comment]
Attached is a hostd file marked up with my commentary for **scenario 1**.

[hostd.log](https://github.com/vmware/vic/files/1472133/hostd.log)

**Comment by sflxn** [comment]
Yesterday, I was able to get pass the commit operation but then it errors on the follow up delete operation.  The cause is the difference between container and handle runtime power state.  After chatting with @hickeng about this subject, he doesn't think we can ever keep them in absolute consistent state.

Today, instead of trying to fix this issue in the portlayer, I tried using simple retries in the persona.  It works well, but I'm seeing a performance problem with the retries.  Currently investigating this.  I've seen this problem in both a physical VC and a VC deployed on Nimbus.

**Comment by sflxn** [comment]
We have a working theory on the performance issue I last posted.  We believe I was maxing out the resource on the ESX host.  I'm submitting a PR and created a new issue for us to return to in the future.  https://github.com/vmware/vic/issues/6785.

For this issue, here and now, I've added retries for the stop and delete in ContainerRm.  We currently already have a test suite for docker rm.  Given this new theory on the performance, it also cast some light on my ability to reproduce this issue.  The environment I was running under was a nested VC cluster with VSAN.  If the only way to reproduce this is to saturate, I'm not sure we should add this stress test to our test.
